### PR TITLE
CORE-19772: Explicitly difference between create and update of registration request

### PR DESCRIPTION
**Pull request by yift-r3**
The issue is that we are trying to persist a new registration request twice (once from the rest worker and once from the membership worker); if both arrive at the same time, we can't find the second one in the database, so hibernate will try to insert both of them and one of the inserts will fail. The worker will recover, but the error will produce an error in the logs, which we can't control.

To solve this, we will persist the request the first time and update it the second time. We will force an error (and retry) if the update request finds no existing entry.

### CODE REVIEW

File: components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/PersistRegistrationRequestHandlerTest.kt [hunk 1]
```diff
@@ -135,7 +138,8 @@ class PersistRegistrationRequestHandlerTest {
     )
 
     private fun getPersistRegistrationRequest(
-        status: RegistrationStatus = RegistrationStatus.PENDING_MEMBER_VERIFICATION
+        status: RegistrationStatus = RegistrationStatus.PENDING_MEMBER_VERIFICATION,
+        create: Boolean = true,
```
Comment (YashNabar) [inline components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/PersistRegistrationRequestHandlerTest.kt]: Considering this is true only for status=`NEW` requests, could we infer this in the persistence handler from the status in the request?

Comment (yift-r3) [reply components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/PersistRegistrationRequestHandlerTest.kt]: Good point. Changed pushed, I'm testing it...

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistRegistrationRequestHandler.kt [hunk 2]
```diff
@@ -22,6 +24,15 @@ internal class PersistRegistrationRequestHandler(
                 registrationId,
                 LockModeType.PESSIMISTIC_WRITE,
             )
+            if ((currentRegistrationRequest == null) &&
+                (request.status != RegistrationStatus.NEW) &&
```
Comment (williamvigorr3) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistRegistrationRequestHandler.kt:28]: Nit: Should we put this `RegistrationStatus.NEW` and `RegistrationStatus.RECEIVED_BY_MGM` in a list?

**Review by YashNabar — DISMISSED** [review]
LGTM (when @williamvigorr3 is happy)

File: components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt [hunk 3]
```diff
@@ -642,12 +642,12 @@ class MembershipPersistenceTest {
     @Test
     fun `serial information can be persisted when requests are processed in unordered manner`() {
```
Comment (williamvigorr3) [inline components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt]: Is this test case still valid now we go from `NEW` to `PENDING_MEMBER_VERIFICATION` (as it's now in order).

File: components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt [hunk 4]
```diff
@@ -1514,18 +1515,22 @@ class MembershipPersistenceTest {
...
         // Persist a completed request
         val registrationId2 = randomUUID().toString()
+        persistRequest(viewOwningHoldingIdentity, registrationId2, RegistrationStatus.RECEIVED_BY_MGM)
         val requestPersistentResult2 =
             persistRequest(viewOwningHoldingIdentity, registrationId2, RegistrationStatus.DECLINED)
         assertThat(requestPersistentResult2).isInstanceOf(MembershipPersistenceResult.Success::class.java)
         // Persist a new request
         val registrationId3 = randomUUID().toString()
+        val holdingId3 = HoldingIdentity(MemberX500Name.parse("O=Charlie, C=GB, L=London"), groupId)
+        persistRequest(holdingId3, registrationId3, RegistrationStatus.NEW)
```
Comment (williamvigorr3) [inline components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt:1539]: I think the comment on line 1528 is wrong. (I was initially confused as to why this is nessary.

**Review by williamvigorr3 — APPROVED** [review]
LGTM

**Pull request merged**